pkg/reconciler/autoscaling/kpa/kpa.go

```diff
@@ -263,13 +270,19 @@ func computeStatus(pa *pav1alpha1.PodAutoscaler, pc podCounts) error {
 	return nil
 }
 
-func reportMetrics(pa *pav1alpha1.PodAutoscaler, pc podCounts) error {
-	serviceLabel := pa.Labels[serving.ServiceLabelKey] // This might be empty.
-	configLabel := pa.Labels[serving.ConfigurationLabelKey]
+func (c *Reconciler) reportMetrics(pa *pav1alpha1.PodAutoscaler, pc podCounts) error {
+	key := pa.Namespace + "+" + pa.Name // `+` is not a valid name char, making strings unique.
+	ctx, ok := c.recorderContextCache.Get(key)
+	if !ok {
+		serviceLabel := pa.Labels[serving.ServiceLabelKey] // This might be empty.
+		configLabel := pa.Labels[serving.ConfigurationLabelKey]
 
-	reportCtx, err := reporterContext(pa.Namespace, serviceLabel, configLabel, pa.Name)
-	if err != nil {
-		return err
+		rctx, err := reporterContext(pa.Namespace, serviceLabel, configLabel, pa.Name)
+		if err != nil {
+			return err
+		}
+		c.recorderContextCache.Add(key, rctx)
```

Comment:
As we'll want this in multiple components: Do we want to plumb this through like this everywhere or could we instead rely on the caches evicting stuff themselves and define the cache globally (and potentially hide it in `reporterContext`) for easy reuse?

Maybe having that `StatsReporter` type wasn't so bad? It'd feel neater to have common `StatsReporter` type that encapsulates this cache handling rather than leaking implementation details of metric reporting (i.e. this cache) to everywhere we might need it 🤔

Reply:
Same here: we have one in Autoscaler and one Activator. So global component probably doesn't make much sense.
But we can iterate depending on how it goes :)